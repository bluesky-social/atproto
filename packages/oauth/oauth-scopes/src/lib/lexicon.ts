export type { LexPermission, LexPermissionSet } from '@atproto/lexicon'
