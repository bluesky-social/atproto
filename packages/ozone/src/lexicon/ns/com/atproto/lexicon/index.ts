/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type Auth,
  type Options as XrpcOptions,
  Server as XrpcServer,
  type StreamConfigOrHandler,
  type MethodConfigOrHandler,
  createServer as createXrpcServer,
} from '@atproto/xrpc-server'
import * as ComAtprotoLexiconSchema from '../../../../types/com/atproto/lexicon/schema.js'
import { Server } from '../../../../index.js'

export class ComAtprotoLexiconNS {
  _server: Server

  constructor(server: Server) {
    this._server = server
  }
}
