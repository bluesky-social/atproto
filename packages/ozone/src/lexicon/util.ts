/**
 * GENERATED CODE - DO NOT MODIFY
 */
export function isObj(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null
}

export function hasProp<K extends PropertyKey>(
  data: object,
  prop: K,
): data is Record<K, unknown> {
  return prop in data
}
