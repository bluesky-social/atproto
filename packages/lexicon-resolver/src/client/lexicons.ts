/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  type LexiconDoc,
  Lexicons,
  ValidationError,
  type ValidationResult,
} from '@atproto/lexicon'
import { type $Typed, is$typed, maybe$typed } from './util.js'

export const schemaDict = {
  ComAtprotoSyncGetRecord: {
    lexicon: 1,
    id: 'com.atproto.sync.getRecord',
    defs: {
      main: {
        type: 'query',
        description:
          'Get data blocks needed to prove the existence or non-existence of record in the current version of repo. Does not require auth.',
        parameters: {
          type: 'params',
          required: ['did', 'collection', 'rkey'],
          properties: {
            did: {
              type: 'string',
              format: 'did',
              description: 'The DID of the repo.',
            },
            collection: {
              type: 'string',
              format: 'nsid',
            },
            rkey: {
              type: 'string',
              description: 'Record Key',
              format: 'record-key',
            },
          },
        },
        output: {
          encoding: 'application/vnd.ipld.car',
        },
        errors: [
          {
            name: 'RecordNotFound',
          },
          {
            name: 'RepoNotFound',
          },
          {
            name: 'RepoTakendown',
          },
          {
            name: 'RepoSuspended',
          },
          {
            name: 'RepoDeactivated',
          },
        ],
      },
    },
  },
} as const satisfies Record<string, LexiconDoc>
export const schemas = Object.values(schemaDict) satisfies LexiconDoc[]
export const lexicons: Lexicons = new Lexicons(schemas)

export function validate<T extends { $type: string }>(
  v: unknown,
  id: string,
  hash: string,
  requiredType: true,
): ValidationResult<T>
export function validate<T extends { $type?: string }>(
  v: unknown,
  id: string,
  hash: string,
  requiredType?: false,
): ValidationResult<T>
export function validate(
  v: unknown,
  id: string,
  hash: string,
  requiredType?: boolean,
): ValidationResult {
  return (requiredType ? is$typed : maybe$typed)(v, id, hash)
    ? lexicons.validate(`${id}#${hash}`, v)
    : {
        success: false,
        error: new ValidationError(
          `Must be an object with "${hash === 'main' ? id : `${id}#${hash}`}" $type property`,
        ),
      }
}

export const ids = {
  ComAtprotoSyncGetRecord: 'com.atproto.sync.getRecord',
} as const
