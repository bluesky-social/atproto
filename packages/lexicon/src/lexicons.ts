import { ZodError } from 'zod'
import {
  LexiconDoc,
  lexiconDoc,
  LexRecord,
  LexXrpcProcedure,
  LexXrpcQuery,
  LexUserType,
  LexiconDocMalformedError,
  LexiconDefNotFoundError,
  InvalidLexiconError,
  ValidationResult,
  ValidationError,
  isObj,
  hasProp,
  LexXrpcSubscription,
} from './types'
import {
  assertValidRecord,
  assertValidXrpcParams,
  assertValidXrpcInput,
  assertValidXrpcOutput,
  assertValidXrpcMessage,
} from './validation'
import { toLexUri } from './util'
import * as ComplexValidators from './validators/complex'

/**
 * A collection of compiled lexicons.
 */
export class Lexicons {
  docs: Map<string, LexiconDoc> = new Map()
  defs: Map<string, LexUserType> = new Map()

  constructor(docs?: unknown[]) {
    if (docs?.length) {
      for (const doc of docs) {
        this.add(doc)
      }
    }
  }

  /**
   * Add a lexicon doc.
   */
  add(doc: unknown): void {
    try {
      lexiconDoc.parse(doc)
    } catch (e) {
      if (e instanceof ZodError) {
        throw new LexiconDocMalformedError(
          `Failed to parse schema definition ${
            (doc as Record<string, string>).id
          }`,
          doc,
          e.issues,
        )
      } else {
        throw e
      }
    }
    const validatedDoc = doc as LexiconDoc
    const uri = toLexUri(validatedDoc.id)
    if (this.docs.has(uri)) {
      throw new Error(`${uri} has already been registered`)
    }

    // WARNING
    // mutates the object
    // -prf
    resolveRefUris(validatedDoc, uri)

    this.docs.set(uri, validatedDoc)
    for (const [defUri, def] of iterDefs(validatedDoc)) {
      this.defs.set(defUri, def)
    }
  }

  /**
   * Remove a lexicon doc.
   */
  remove(uri: string) {
    uri = toLexUri(uri)
    const doc = this.docs.get(uri)
    if (!doc) {
      throw new Error(`Unable to remove "${uri}": does not exist`)
    }
    for (const [defUri, _def] of iterDefs(doc)) {
      this.defs.delete(defUri)
    }
    this.docs.delete(uri)
  }

  /**
   * Get a lexicon doc.
   */
  get(uri: string): LexiconDoc | undefined {
    uri = toLexUri(uri)
    return this.docs.get(uri)
  }

  /**
   * Get a definition.
   */
  getDef(uri: string): LexUserType | undefined {
    uri = toLexUri(uri)
    return this.defs.get(uri)
  }

  /**
   * Get a def, throw if not found. Throws on not found.
   */
  getDefOrThrow(uri: string, types?: string[]): LexUserType {
    const def = this.getDef(uri)
    if (!def) {
      throw new LexiconDefNotFoundError(`Lexicon not found: ${uri}`)
    }
    if (types && !types.includes(def.type)) {
      throw new InvalidLexiconError(
        `Not a ${types.join(' or ')} lexicon: ${uri}`,
      )
    }
    return def
  }

  /**
   * Validate a record or object.
   */
  validate(lexUri: string, value: unknown): ValidationResult {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, ['record', 'object'])
    if (!isObj(value)) {
      throw new ValidationError(`Value must be an object`)
    }
    if (def.type === 'record') {
      return ComplexValidators.object(this, 'Record', def.record, value)
    } else if (def.type === 'object') {
      return ComplexValidators.object(this, 'Object', def, value)
    } else {
      // shouldnt happen
      throw new InvalidLexiconError('Definition must be a record or object')
    }
  }

  /**
   * Validate a record and throw on any error.
   */
  assertValidRecord(lexUri: string, value: unknown) {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, ['record'])
    if (!isObj(value)) {
      throw new ValidationError(`Record must be an object`)
    }
    if (!hasProp(value, '$type') || typeof value.$type !== 'string') {
      throw new ValidationError(`Record/$type must be a string`)
    }
    const $type = (value as Record<string, string>).$type || ''
    if (toLexUri($type) !== lexUri) {
      throw new ValidationError(
        `Invalid $type: must be ${lexUri}, got ${$type}`,
      )
    }
    return assertValidRecord(this, def as LexRecord, value)
  }

  /**
   * Validate xrpc query params and throw on any error.
   */
  assertValidXrpcParams(lexUri: string, value: unknown) {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, [
      'query',
      'procedure',
      'subscription',
    ])
    return assertValidXrpcParams(
      this,
      def as LexXrpcProcedure | LexXrpcQuery | LexXrpcSubscription,
      value,
    )
  }

  /**
   * Validate xrpc input body and throw on any error.
   */
  assertValidXrpcInput(lexUri: string, value: unknown) {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, ['procedure'])
    return assertValidXrpcInput(this, def as LexXrpcProcedure, value)
  }

  /**
   * Validate xrpc output body and throw on any error.
   */
  assertValidXrpcOutput(lexUri: string, value: unknown) {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, ['query', 'procedure'])
    return assertValidXrpcOutput(
      this,
      def as LexXrpcProcedure | LexXrpcQuery,
      value,
    )
  }

  /**
   * Validate xrpc subscription message and throw on any error.
   */
  assertValidXrpcMessage<T = unknown>(lexUri: string, value: unknown): T {
    lexUri = toLexUri(lexUri)
    const def = this.getDefOrThrow(lexUri, ['subscription'])
    return assertValidXrpcMessage(this, def as LexXrpcSubscription, value) as T
  }

  /**
   * Resolve a lex uri given a ref
   */
  resolveLexUri(lexUri: string, ref: string) {
    lexUri = toLexUri(lexUri)
    return toLexUri(ref, lexUri)
  }
}

function* iterDefs(doc: LexiconDoc): Generator<[string, LexUserType]> {
  for (const defId in doc.defs) {
    yield [`lex:${doc.id}#${defId}`, doc.defs[defId]]
    if (defId === 'main') {
      yield [`lex:${doc.id}`, doc.defs[defId]]
    }
  }
}

// WARNING
// this method mutates objects
// -prf
function resolveRefUris(obj: any, baseUri: string): any {
  for (const k in obj) {
    if (obj.type === 'ref') {
      obj.ref = toLexUri(obj.ref, baseUri)
    } else if (obj.type === 'union') {
      obj.refs = obj.refs.map((ref) => toLexUri(ref, baseUri))
    } else if (Array.isArray(obj[k])) {
      obj[k] = obj[k].map((item: any) => {
        if (typeof item === 'string') {
          return item.startsWith('#') ? toLexUri(item, baseUri) : item
        } else if (item && typeof item === 'object') {
          return resolveRefUris(item, baseUri)
        }
        return item
      })
    } else if (obj[k] && typeof obj[k] === 'object') {
      obj[k] = resolveRefUris(obj[k], baseUri)
    }
  }
  return obj
}
