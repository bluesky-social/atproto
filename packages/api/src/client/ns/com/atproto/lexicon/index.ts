/**
 * GENERATED CODE - DO NOT MODIFY
 */
import {
  XrpcClient,
  type FetchHandler,
  type FetchHandlerOptions,
} from '@atproto/xrpc'
import { schemas } from '../../../../lexicons.js'
import { CID } from 'multiformats/cid'
import { type OmitKey, type Un$Typed } from '../../../../util.js'
import * as ComAtprotoLexiconSchema from '../../../../types/com/atproto/lexicon/schema.js'
import * as ComAtprotoRepoCreateRecord from '../../../../types/com/atproto/repo/createRecord.js'
import * as ComAtprotoRepoDeleteRecord from '../../../../types/com/atproto/repo/deleteRecord.js'
import * as ComAtprotoRepoGetRecord from '../../../../types/com/atproto/repo/getRecord.js'
import * as ComAtprotoRepoListRecords from '../../../../types/com/atproto/repo/listRecords.js'
import * as ComAtprotoRepoPutRecord from '../../../../types/com/atproto/repo/putRecord.js'

export class ComAtprotoLexiconNS {
  _client: XrpcClient
  schema: ComAtprotoLexiconSchemaRecord

  constructor(client: XrpcClient) {
    this._client = client
    this.schema = new ComAtprotoLexiconSchemaRecord(client)
  }
}

export class ComAtprotoLexiconSchemaRecord {
  _client: XrpcClient

  constructor(client: XrpcClient) {
    this._client = client
  }

  async list(
    params: OmitKey<ComAtprotoRepoListRecords.QueryParams, 'collection'>,
  ): Promise<{
    cursor?: string
    records: { uri: string; value: ComAtprotoLexiconSchema.Record }[]
  }> {
    const res = await this._client.call('com.atproto.repo.listRecords', {
      collection: 'com.atproto.lexicon.schema',
      ...params,
    })
    return res.data
  }

  async get(
    params: OmitKey<ComAtprotoRepoGetRecord.QueryParams, 'collection'>,
  ): Promise<{
    uri: string
    cid: string
    value: ComAtprotoLexiconSchema.Record
  }> {
    const res = await this._client.call('com.atproto.repo.getRecord', {
      collection: 'com.atproto.lexicon.schema',
      ...params,
    })
    return res.data
  }

  async create(
    params: OmitKey<
      ComAtprotoRepoCreateRecord.InputSchema,
      'collection' | 'record'
    >,
    record: Un$Typed<ComAtprotoLexiconSchema.Record>,
    headers?: Record<string, string>,
  ): Promise<{ uri: string; cid: string }> {
    const collection = 'com.atproto.lexicon.schema'
    const res = await this._client.call(
      'com.atproto.repo.createRecord',
      undefined,
      { collection, ...params, record: { ...record, $type: collection } },
      { encoding: 'application/json', headers },
    )
    return res.data
  }

  async put(
    params: OmitKey<
      ComAtprotoRepoPutRecord.InputSchema,
      'collection' | 'record'
    >,
    record: Un$Typed<ComAtprotoLexiconSchema.Record>,
    headers?: Record<string, string>,
  ): Promise<{ uri: string; cid: string }> {
    const collection = 'com.atproto.lexicon.schema'
    const res = await this._client.call(
      'com.atproto.repo.putRecord',
      undefined,
      { collection, ...params, record: { ...record, $type: collection } },
      { encoding: 'application/json', headers },
    )
    return res.data
  }

  async delete(
    params: OmitKey<ComAtprotoRepoDeleteRecord.InputSchema, 'collection'>,
    headers?: Record<string, string>,
  ): Promise<void> {
    await this._client.call(
      'com.atproto.repo.deleteRecord',
      undefined,
      { collection: 'com.atproto.lexicon.schema', ...params },
      { headers },
    )
  }
}
