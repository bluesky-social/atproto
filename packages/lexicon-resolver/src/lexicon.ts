import dns from 'node:dns/promises'
import { CID } from 'multiformats/cid'
import { LexiconDoc, parseLexiconDoc } from '@atproto/lexicon'
import { Commit } from '@atproto/repo'
import { AtUri, NSID, ensureValidDid } from '@atproto/syntax'
import {
  BuildRecordResolverOptions,
  ResolveRecordOptions,
  buildRecordResolver,
} from './record.js'
import { isValidDid } from './util.js'

const DNS_SUBDOMAIN = '_lexicon'
const DNS_PREFIX = 'did='
export const LEXICON_SCHEMA_NSID = 'com.atproto.lexicon.schema'

/**
 * Resolve Lexicon from an NSID
 */
export type LexiconResolver = (
  nsid: NSID | string,
) => Promise<LexiconResolution>

/**
 * Resolve Lexicon from an NSID using Lexicon DID authority and record resolution
 */
export type AtprotoLexiconResolver = (
  nsid: NSID | string,
  options?: ResolveLexiconOptions,
) => Promise<LexiconResolution>

export type BuildLexiconResolverOptions = BuildRecordResolverOptions

export type ResolveLexiconOptions = ResolveRecordOptions & {
  didAuthority?: string
}

export type LexiconResolution = {
  commit: Commit
  uri: AtUri
  cid: CID
  nsid: NSID
  lexicon: LexiconDoc & LexiconSchemaRecord
}

/**
 * Build a Lexicon resolver function.
 */
export function buildLexiconResolver(
  options: BuildLexiconResolverOptions = {},
): AtprotoLexiconResolver {
  const resolveRecord = buildRecordResolver(options)
  return async function (
    nsidStr: NSID | string,
    opts: ResolveLexiconOptions = {},
  ): Promise<LexiconResolution> {
    const nsid = typeof nsidStr === 'string' ? NSID.parse(nsidStr) : nsidStr
    const didAuthority = await getDidAuthority(nsid, opts)
    const verified = await resolveRecord(
      AtUri.make(didAuthority, LEXICON_SCHEMA_NSID, nsid.toString()),
      { forceRefresh: opts.forceRefresh },
    ).catch((err) => {
      throw new LexiconResolutionError(
        'Could not resolve Lexicon schema record',
        { cause: err },
      )
    })
    let lexicon: LexiconDoc
    try {
      lexicon = parseLexiconDoc(verified.record)
    } catch (err) {
      throw new LexiconResolutionError('Invalid Lexicon document', {
        cause: err,
      })
    }
    if (!isLexiconSchemaRecord(lexicon)) {
      throw new LexiconResolutionError('Invalid Lexicon schema record')
    }
    if (lexicon.id !== nsid.toString()) {
      throw new LexiconResolutionError(
        `Lexicon schema record id does not match NSID: ${lexicon.id}`,
      )
    }
    const { uri, cid, commit } = verified
    return { commit, uri, cid, nsid, lexicon }
  } satisfies LexiconResolver
}

export const resolveLexicon = buildLexiconResolver()

/**
 * Resolve the DID authority for a Lexicon from the network using DNS, based on its NSID.
 * @param nsidStr NSID or string representing one for which to lookup its Lexicon DID authority.
 */
export async function resolveLexiconDidAuthority(
  nsidStr: NSID | string,
): Promise<string | undefined> {
  const nsid = typeof nsidStr === 'string' ? NSID.parse(nsidStr) : nsidStr
  const did = await resolveDns(nsid.authority)
  if (did == null || !isValidDid(did)) return
  return did
}

export class LexiconResolutionError extends Error {
  constructor(message?: string, options?: ErrorOptions) {
    super(message, options)
    this.name = 'LexiconResolutionError'
  }
}

async function getDidAuthority(nsid: NSID, options: ResolveLexiconOptions) {
  if (options.didAuthority) {
    ensureValidDid(options.didAuthority)
    return options.didAuthority
  }
  const did = await resolveLexiconDidAuthority(nsid)
  if (!did) {
    throw new LexiconResolutionError(
      `Could not resolve a DID authority for NSID: ${nsid}`,
    )
  }
  return did
}

async function resolveDns(authority: string): Promise<string | undefined> {
  let chunkedResults: string[][]
  try {
    chunkedResults = await dns.resolveTxt(`${DNS_SUBDOMAIN}.${authority}`)
  } catch (err) {
    return undefined
  }
  return parseDnsResult(chunkedResults)
}

function parseDnsResult(chunkedResults: string[][]): string | undefined {
  const results = chunkedResults.map((chunks) => chunks.join(''))
  const found = results.filter((i) => i.startsWith(DNS_PREFIX))
  if (found.length !== 1) {
    return undefined
  }
  return found[0].slice(DNS_PREFIX.length)
}

type LexiconSchemaRecord = { $type: typeof LEXICON_SCHEMA_NSID }
function isLexiconSchemaRecord(v: unknown): v is LexiconSchemaRecord {
  return v?.['$type'] === LEXICON_SCHEMA_NSID
}
