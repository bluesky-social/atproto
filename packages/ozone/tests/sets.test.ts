import { TestNetwork, SeedClient, basicSeed } from '@atproto/dev-env'
import AtpAgent, {
  ToolsOzoneSetDefs,
  ToolsOzoneSetQuerySets,
} from '@atproto/api'
import { forSnapshot } from './_util'
import { ids } from '../src/lexicon/lexicons'

describe('ozone-sets', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let sc: SeedClient

  const sampleSet1 = {
    name: 'test-set-1',
    description: 'Test set 1',
  }

  const sampleSet2 = {
    name: 'test-set-2',
  }

  const sampleSet3 = {
    name: 'another-set',
    description: 'Another test set',
  }

  const upsertSet = async (set: ToolsOzoneSetDefs.Set) => {
    const { data } = await agent.tools.ozone.set.upsertSet(set, {
      encoding: 'application/json',
      headers: await network.ozone.modHeaders(
        ids.ToolsOzoneSetUpsertSet,
        'admin',
      ),
    })

    return data
  }

  const removeSet = async (name: string) => {
    await agent.tools.ozone.set.deleteSet(
      { name },
      {
        encoding: 'application/json',
        headers: await network.ozone.modHeaders(
          ids.ToolsOzoneSetDeleteSet,
          'admin',
        ),
      },
    )
  }

  const addValues = async (name: string, values: string[]) => {
    await agent.tools.ozone.set.addValues(
      { name, values },
      {
        encoding: 'application/json',
        headers: await network.ozone.modHeaders(
          ids.ToolsOzoneSetAddValues,
          'admin',
        ),
      },
    )
  }

  const getValues = async (name: string, limit?: number, cursor?: string) => {
    const { data } = await agent.tools.ozone.set.getValues(
      { name, limit, cursor },
      {
        headers: await network.ozone.modHeaders(
          ids.ToolsOzoneSetGetValues,
          'moderator',
        ),
      },
    )
    return data
  }

  const querySets = async (params: ToolsOzoneSetQuerySets.QueryParams) => {
    const { data } = await agent.tools.ozone.set.querySets(params, {
      headers: await network.ozone.modHeaders(
        ids.ToolsOzoneSetQuerySets,
        'moderator',
      ),
    })
    return data
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'ozone_sets',
    })
    agent = network.ozone.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await network.processAll()
  })

  afterAll(async () => {
    await network.close()
  })

  describe('querySets', () => {
    beforeAll(async () => {
      await Promise.all([
        upsertSet(sampleSet1),
        upsertSet(sampleSet2),
        upsertSet(sampleSet3),
      ])
    })
    afterAll(async () => {
      await Promise.all([
        removeSet(sampleSet1.name),
        removeSet(sampleSet2.name),
        removeSet(sampleSet3.name),
      ])
    })
    it('returns all sets when no parameters are provided', async () => {
      const result = await querySets({})
      expect(result.sets.length).toBe(3)
      expect(forSnapshot(result.sets)).toMatchSnapshot()
    })

    it('limits the number of returned sets', async () => {
      const result = await querySets({ limit: 2 })
      expect(result.sets.length).toBe(2)
      expect(result.cursor).toBeDefined()
    })

    it('returns sets after the cursor', async () => {
      const firstPage = await querySets({ limit: 2 })
      const secondPage = await querySets({ cursor: firstPage.cursor })
      expect(secondPage.sets.length).toBe(1)
      expect(secondPage.sets[0].name).toBe('test-set-2')
    })

    it('filters sets by name prefix', async () => {
      const result = await querySets({ namePrefix: 'test-' })
      expect(result.sets.length).toBe(2)
      expect(result.sets.map((s) => s.name)).toEqual([
        'test-set-1',
        'test-set-2',
      ])
    })

    it('sorts sets by given column and direction', async () => {
      const sortedByName = await querySets({ sortBy: 'name' })
      expect(sortedByName.sets.map((s) => s.name)).toEqual([
        'another-set',
        'test-set-1',
        'test-set-2',
      ])
      const reverseSortedByName = await querySets({
        sortBy: 'name',
        sortDirection: 'desc',
      })
      expect(reverseSortedByName.sets.map((s) => s.name)).toEqual([
        'test-set-2',
        'test-set-1',
        'another-set',
      ])
    })
  })

  describe('upsertSet', () => {
    afterAll(async () => {
      await removeSet('new-test-set')
    })
    it('creates a new set', async () => {
      const result = await upsertSet({
        name: 'new-test-set',
        description: 'A new test set',
      })
      expect(forSnapshot(result)).toMatchSnapshot()
    })

    it('updates an existing set', async () => {
      const result = await upsertSet({
        name: 'new-test-set',
        description: 'Updated description',
      })
      expect(forSnapshot(result)).toMatchSnapshot()
    })

    it('allows setting empty description', async () => {
      const result = await upsertSet({
        name: 'new-test-set',
        description: '',
      })
      expect(result.description).toBeUndefined()
    })
  })

  describe('addValues', () => {
    beforeAll(async () => {
      await upsertSet(sampleSet1)
      await upsertSet(sampleSet2)
    })
    afterAll(async () => {
      await removeSet(sampleSet1.name)
      await removeSet(sampleSet2.name)
    })
    it('adds new values to an existing set', async () => {
      const newValues = ['value1', 'value2', 'value3']
      await addValues(sampleSet1.name, newValues)

      const result = await getValues(sampleSet1.name)
      expect(result.values).toEqual(expect.arrayContaining(newValues))
    })

    it('does not duplicate existing values', async () => {
      const initialValues = ['initial1', 'initial2']
      await addValues(sampleSet2.name, initialValues)

      const newValues = ['initial2', 'new1', 'new2']
      await addValues(sampleSet2.name, newValues)

      const result = await getValues(sampleSet2.name)
      expect(result.values).toEqual(
        expect.arrayContaining([...initialValues, 'new1', 'new2']),
      )
      expect(result.values.filter((v) => v === 'initial2').length).toBe(1)
    })
  })

  describe('getValues', () => {
    beforeAll(async () => {
      await upsertSet(sampleSet1)
    })
    afterAll(async () => {
      await removeSet(sampleSet1.name)
    })
    it('paginates values from a set', async () => {
      const allValues = Array.from({ length: 9 }, (_, i) => `value${i}`)
      await addValues(sampleSet1.name, allValues)

      const firstPage = await getValues(sampleSet1.name, 3)
      const secondPage = await getValues(sampleSet1.name, 3, firstPage.cursor)
      const lastPage = await getValues(sampleSet1.name, 3, secondPage.cursor)

      expect(firstPage.values).toEqual(allValues.slice(0, 3))
      expect(secondPage.values).toEqual(allValues.slice(3, 6))
      expect(lastPage.values).toEqual(allValues.slice(6, 9))
    })
  })
})
