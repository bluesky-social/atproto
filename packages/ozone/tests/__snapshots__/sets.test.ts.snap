// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ozone-sets querySets returns all sets when no parameters are provided 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "description": "Another test set",
    "name": "another-set",
    "setSize": 0,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "description": "Test set 1",
    "name": "test-set-1",
    "setSize": 0,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "name": "test-set-2",
    "setSize": 0,
    "updatedAt": "1970-01-01T00:00:00.000Z",
  },
]
`;

exports[`ozone-sets upsertSet creates a new set 1`] = `
Object {
  "createdAt": "1970-01-01T00:00:00.000Z",
  "description": "A new test set",
  "name": "new-test-set",
  "setSize": 0,
  "updatedAt": "1970-01-01T00:00:00.000Z",
}
`;

exports[`ozone-sets upsertSet updates an existing set 1`] = `
Object {
  "createdAt": "1970-01-01T00:00:00.000Z",
  "description": "Updated description",
  "name": "new-test-set",
  "setSize": 0,
  "updatedAt": "1970-01-01T00:00:00.000Z",
}
`;
